Make DDPOptimizer work with torch._dynamo.explain()

wconstab opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #94749

GraphModules that were created during DDPOptimizer graph breaking
lacked `compile_subgraph_reason`, which caused an exception when
running .explain().

Now the reason is provided and users can use .explain() to find out
that DDPOptimizer is causing graph breaks.

Fixes #94579

cc @mlazos @soumith @voznesenskym @yanboliang @penguinwu @anijain2305 @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @Xia-Weiwen @wenzhe-nrv @jiayisunx @desertfire

wconstab committed:
Make DDPOptimizer work with torch._dynamo.explain()

GraphModules that were created during DDPOptimizer graph breaking
lacked `compile_subgraph_reason`, which caused an exception when
running .explain().

Now the reason is provided and users can use .explain() to find out
that DDPOptimizer is causing graph breaks.

Fixes #94579

[ghstack-poisoned]

voznesenskym commented on test/distributed/test_dynamo_distributed.py:
```diff
@@ -399,6 +399,11 @@ def opt_fn(inputs):
         self.assertEqual(check_splits_compiler.compiler_called, 3)
 
+        # ensure compatibilty with dynamo explain
+        _, _, _, _, break_reasons, _ = torch._dynamo.explain(ddp_m, inputs)
+        self.assertEqual(len(break_reasons), 3)
```
nit, this might be a bit saner: 
```suggestion
        explain_out = torch._dynamo.explain(ddp_m, inputs)
        break_reasons = explain_out[4]
        self.assertEqual(len(break_reasons), 3)
```

wconstab committed:
Update on "Make DDPOptimizer work with torch._dynamo.explain()"


GraphModules that were created during DDPOptimizer graph breaking
lacked `compile_subgraph_reason`, which caused an exception when
running .explain().

Now the reason is provided and users can use .explain() to find out
that DDPOptimizer is causing graph breaks.

Fixes #94579

cc mlazos soumith voznesenskym yanboliang penguinwu anijain2305 EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx desertfire

[ghstack-poisoned]

wconstab commented on torch/_dynamo/backends/distributed.py:
```diff
@@ -259,6 +261,14 @@ def forward(self, *args):
                 input_mod.recompile()
+                input_mod.compile_subgraph_reason = GraphCompileReason(
+                    "DDPOptimizer intentional graph-break to improve communication overlap."
+                    " Set `torch._dynamo.config.optimize_ddp = False` to disable",
```
yea just added a pointer. good idea.

wconstab commented:
@pytorchbot merge

wconstab committed:
Update on "Make DDPOptimizer work with torch._dynamo.explain()"


GraphModules that were created during DDPOptimizer graph breaking
lacked `compile_subgraph_reason`, which caused an exception when
running .explain().

Now the reason is provided and users can use .explain() to find out
that DDPOptimizer is causing graph breaks.

Fixes #94579

cc mlazos soumith voznesenskym yanboliang penguinwu anijain2305 EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx desertfire

[ghstack-poisoned]

wconstab commented:
@pytorchbot merge

wconstab commented:
@pytorchbot merge -f "Broken trunk"
